Firestore/Example/Tests/Integration/FSTTransactionTests.mm

```diff
@@ -20,184 +20,372 @@
-  FIRDocumentReference *doc = [[firestore collectionWithPath:@"towns"] documentWithAutoID];
-  [self writeDocumentRef:doc data:@{@"foo" : @"bar"}];
-  FIRDocumentSnapshot *snapshot = [self readDocumentForRef:doc];
-  XCTAssertEqualObjects(@"bar", snapshot[@"foo"]);
+- (void)assertNilError:(NSError *)error message:(NSString *)message {
+  XCTAssertNil(error, @"%@", message);
+}
 
-  XCTestExpectation *expectation = [self expectationWithDescription:@"transaction"];
-  [firestore
+- (void)assertError:(NSError *)error message:(NSString *)message {
+  XCTAssertNotNil(error, @"%@", message);
+}
+
+- (void)assertSnapshot:(FIRDocumentSnapshot *)snapshot
+          equalsObject:(NSObject *)expected
+                 error:(NSError *)error {
+  XCTAssertNil(error);
+  XCTAssertTrue(snapshot.exists);
+  XCTAssertEqualObjects(expected, snapshot.data);
+}
+@end
+
+typedef void (^TransactionStage)(FIRTransaction *, FIRDocumentReference *);
+
+/**
+ * The transaction stages that follow are postfixed by numbers to indicate the calling order. For
+ * example, calling `set1` followed by `set2` should result in the document being set to the value
+ * specified by `set2`.
+ */
+TransactionStage delete1 = ^(FIRTransaction *transaction, FIRDocumentReference *doc) {
+  [transaction deleteDocument:doc];
+};
+
+TransactionStage update1 = ^(FIRTransaction *transaction, FIRDocumentReference *doc) {
+  [transaction updateData:@{@"foo" : @"bar1"} forDocument:doc];
+};
+
+TransactionStage update2 = ^(FIRTransaction *transaction, FIRDocumentReference *doc) {
+  [transaction updateData:@{@"foo" : @"bar2"} forDocument:doc];
+};
+
+TransactionStage set1 = ^(FIRTransaction *transaction, FIRDocumentReference *doc) {
+  [transaction setData:@{@"foo" : @"bar1"} forDocument:doc];
+};
+
+TransactionStage set2 = ^(FIRTransaction *transaction, FIRDocumentReference *doc) {
+  [transaction setData:@{@"foo" : @"bar2"} forDocument:doc];
+};
+
+TransactionStage get = ^(FIRTransaction *transaction, FIRDocumentReference *doc) {
+  NSError *error = nil;
+  [transaction getDocument:doc error:&error];
+};
+
+/**
+ * Used for testing that all possible combinations of executing transactions result in the desired
+ * document value or error.
+ *
+ * `runWithStages`, `withExistingDoc`, and `withNonexistentDoc` don't actually do anything except
+ * assign variables into `FSTTransactionTester`.
+ *
+ * `expectDoc`, `expectNoDoc`, and `expectError` will trigger the transaction to run and assert
+ * that the end result matches the input.
+ */
+@interface FSTTransactionTester : NSObject
+- (FSTTransactionTester *)withExistingDoc;
+- (FSTTransactionTester *)withNonexistentDoc;
+- (FSTTransactionTester *)runWithStages:(NSArray<TransactionStage> *)stages;
+- (void)expectDoc:(NSObject *)expected;
+- (void)expectNoDoc;
+- (void)expectError:(FIRFirestoreErrorCode)expected;
+@end
+
+@implementation FSTTransactionTester {
+  FIRFirestore *_db;
+  FIRDocumentReference *_docRef;
+  boolean_t _fromExistingDoc;
```

Comment:
Use `BOOL` for Objective-C booleans or `bool` for C++. `boolean_t` is a kernel type we should never use.